### PR TITLE
imudp: proposed fix for issue #2040 - avoid socket descriptor leak while binding to a nonlocal address

### PR DESCRIPTION
**Pull request by amg1127**
https://github.com/rsyslog/rsyslog/issues/2040

### CODE REVIEW

File: runtime/net.c [hunk 1]
```diff
@@ -1485,13 +1485,15 @@ create_udp_socket(uchar *hostname,
 						if (ipfreebind >= IPFREEBIND_ENABLED_WITH_LOG)
 							errmsg.LogMsg(0, RS_RET_OK_WARN, LOG_WARNING,
 								"bound address %s IP free", hostname);
-						continue;
+						goto bindsuccess;
```
Comment (rgerhards) [inline runtime/net.c]: We should not use ```goto``` other than to exit a function (via it's finalizer).

**Comment by amg1127** [comment]
@rgerhards I was able to remove the `goto` statement while maintaining the original handling of _EADDRNOTAVAIL_ error.

**Comment by rgerhards** [comment]
I admit I am not happy with the solution- looks a bit hackish to me, introduces a kind of goto instead of looking at the actually desired code structure. On the other hand, the original code looks hackish at all, it's probably one of the darkest and oldest pieces inside rsyslog. What makes matters worse is that IPFREEBIND feature cannot be configured, but is hard coded. So if we patch this now, we introduce a change of behaviour. Well, well, this looks like something I really need to check on how to best clean up :-(

Sorry for the hassle, I should have noticed that when I merged the original change for IPFREEDBIND...

**Comment by rgerhards** [comment]
oh oh, this **really** needs a refactoring...

**Comment by rgerhards** [comment]
> What makes matters worse is that IPFREEBIND feature cannot be configured, but is hard coded.

I must correct myself: it's just that there are some definitions, but integer values are used ;-) So the good news is that it is configurable.

**Comment by rgerhards** [comment]
I now have used your original patch (idea) to fix this in the refactored code, see #2080. Thanks a ton for all your help! You could not do this cleaner than you did without refactoring...

**Comment by rgerhards** [comment]
closed via https://github.com/rsyslog/rsyslog/pull/2080

**Comment by amg1127** [comment]
The hassle is not a problem at all. I enjoy your desire for keeping good coding standards into the software and am aware that my fix proposal was not as good as required.

I am used to coding small softwares for hobby, however I have been working as an IT system administrator for a long time, so I would not be able to study all "_rsyslog_" code and write a good patch in a timely manner. I opted to change the code as less as possible instead.

Thank you for solving my issue and restoring the IPFREEBIND feature. Enabling the "_ip_nonlocal_bind_" kernel option on the OS would let me a little annoyed, as it would allow all processes to bind nonlocal addresses. I am looking forward for the next stable RPM package.

**Pull request closed without merging**